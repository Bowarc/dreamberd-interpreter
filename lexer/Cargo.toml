[package]
name = "lexer"
version = "0.1.0"
edition = "2021"

[dependencies]
thiserror = "2.0.11"
enum_variant_name = { git = "https://github.com/bowarc/crates", package = "enum_variant_name" }

[dev-dependencies]
insta = "1.42.1"
