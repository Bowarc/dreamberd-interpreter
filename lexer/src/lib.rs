mod error;
mod token;

pub use error::LexerError;
pub use token::Token;

pub fn scan<T: std::io::Read>(input: T) -> Result<Vec<Token>, LexerError> {
    use std::{
        fmt::Write as _,
        io::{BufRead as _, BufReader},
    };
    let mut reader = BufReader::new(input);

    let mut tokens = Vec::new();

    loop {
        let mut line = String::new();

        let read = reader.read_line(&mut line).unwrap();
        if read == 0 {
            break;
        }

        println!("Read {read}bytes");

        let line_tokens = match scan_one(&line) {
            Ok(tokens) => {
                println!(
                    "{}",
                    tokens.iter().fold(String::new(), |mut output, t| {
                        let _ = writeln!(output, "  {t:?}");
                        output
                    })
                );
                tokens
            }
            Err(e) => {
                eprintln!("{e}");
                break;
            }
        };

        tokens.extend(line_tokens);
    }

    tokens.push(Token::EOF);

    Ok(tokens)
}

// In a normal language i would handle the strings here, like, actually parsing "hi" as a single token
// But in dreamberd, """hi""" is a valid string, same as '''''hi''''', or even hi
// coupled with the fact that you can redefine keywords,
// i have no way to parse a string from just one line of context
pub fn scan_one(line: &str) -> Result<Vec<Token>, LexerError> {
    use Token;

    let mut tokens = Vec::new();

    for c in line.chars() {
        // Basic single character immutable tokens like '"!, etc..
        if let Ok(token) = Token::try_from(c) {
            tokens.push(token);
            continue;
        }

        // Handling numbers
        if c.is_ascii_digit() {
            // Here we remove 48 when converting to an u8, to keep the numerical value
            // (instead of getting the unicode value)

            if let Some(Token::Litteral(inner)) = tokens.last_mut() {
                inner.push(c);
            } else if let Some(Token::Numeric(inner)) = tokens.last_mut() {
                inner.push(c as u8 - 48);
            } else {
                tokens.push(Token::Numeric(vec![c as u8 - 48]));
            }
            continue;
        }

        // Everything else is a litteral for the lexer
        if let Some(Token::Litteral(inner)) = tokens.last_mut() {
            inner.push(c);
        } else {
            tokens.push(Token::Litteral(c.to_string()))
        }
    }

    Ok(tokens)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_single_character_tokens() {
        assert_debug_snapshot!(scan_one("."));
        assert_debug_snapshot!(scan_one("!"));
        assert_debug_snapshot!(scan_one("="));
        assert_debug_snapshot!(scan_one(","));
        assert_debug_snapshot!(scan_one("["));
        assert_debug_snapshot!(scan_one("]"));
        assert_debug_snapshot!(scan_one("("));
        assert_debug_snapshot!(scan_one(")"));
        assert_debug_snapshot!(scan_one("'"));
        assert_debug_snapshot!(scan_one("\""));
        assert_debug_snapshot!(scan_one("<"));
        assert_debug_snapshot!(scan_one(">"));
        assert_debug_snapshot!(scan_one("+"));
        assert_debug_snapshot!(scan_one("-"));
        assert_debug_snapshot!(scan_one("/"));
        assert_debug_snapshot!(scan_one("*"));
    }

    #[test]
    fn test_numeric_tokens() {
        assert_debug_snapshot!(scan_one("123"));
        assert_debug_snapshot!(scan_one("4567"));
        assert_debug_snapshot!(scan_one("0"));
    }

    #[test]
    fn test_literal_tokens() {
        assert_debug_snapshot!(scan_one("hello"));
        assert_debug_snapshot!(scan_one("world"));
        assert_debug_snapshot!(scan_one("foo bar"));
    }

    #[test]
    fn test_mixed_tokens() {
        assert_debug_snapshot!(scan_one("const var scores = [3, 2, 5]!"));
        assert_debug_snapshot!(scan_one("const x = 10 + 20!"));
        assert_debug_snapshot!(scan_one("if (x < 5) { return! }"));
    }

    #[test]
    fn test_edge_cases() {
        assert_debug_snapshot!(scan_one(""));

        assert_debug_snapshot!(scan_one("   "));
        assert_debug_snapshot!(scan_one("\t\t\t"));
        assert_debug_snapshot!(scan_one("\n\n\n"));

        assert_debug_snapshot!(scan_one("@#$%^&*()"));

        assert_debug_snapshot!(scan_one(&"a".repeat(1000)));

        assert_debug_snapshot!(scan_one("!!!"));
        assert_debug_snapshot!(scan_one("..."));
        assert_debug_snapshot!(scan_one(",,,"));

        assert_debug_snapshot!(scan_one("123abc!@#"));

        assert_debug_snapshot!(scan_one("[(])"));
        assert_debug_snapshot!(scan_one("[{]}"));

        assert_debug_snapshot!(scan_one("\"Hello\\nWorld\""));

        assert_debug_snapshot!(scan_one("00123"));
        assert_debug_snapshot!(scan_one("0001"));

        assert_debug_snapshot!(scan_one("Hello\x00World"));

        assert_debug_snapshot!(scan_one("12345678901234567890"));
    }
}

// small test
// match &tokens[..5] {
//     [outer, Token::Space, inner, Token::Space, Token::Litteral(var_name)] => {
//         println!(
//             "Declaration of variable '{var_name}' with{} outer mutablility and with{} inner mutability",
//             if outer == &Token::Litteral(String::from("const")) {
//                 "out"
//             } else {
//                 ""
//             },
//             if inner == &Token::Litteral(String::from("const")) {
//                 "out"
//             } else {
//                 ""
//             }
//         )
//     }
//     a => {
//         println!("{a:?}")
//     }
// }
