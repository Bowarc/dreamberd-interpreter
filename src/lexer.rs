mod error;
mod token;

pub fn scan(source: super::SourceFile) -> Result<Vec<token::Token>, error::ParserError> {
    use std::{
        fmt::Write as _,
        fs::OpenOptions,
        io::{BufRead as _, BufReader},
    };

    let mut tokens = Vec::new();

    let Ok(file) = OpenOptions::new().read(true).open(source.0) else {
        return Err(error::ParserError::SourceFileNotFound);
    };

    let mut reader = BufReader::new(file);

    loop {
        let mut line = String::new();

        let read = reader.read_line(&mut line).unwrap();
        if read == 0 {
            break;
        }

        println!("Read {read}bytes");

        tokens = match scan_one(&line) {
            Ok(tokens) => {
                println!(
                    "{}",
                    tokens.iter().fold(String::new(), |mut output, t| {
                        let _ = writeln!(output, "  {t:?}");
                        output
                    })
                );
                tokens
            }
            Err(e) => {
                eprintln!("{e}");
                break;
            }
        };
    }

    Ok(tokens)
}

fn scan_one(line: &str) -> Result<Vec<token::Token>, error::ParserError> {
    use token::Token;

    let mut tokens = Vec::new();

    for c in line.chars() {
        match c {
            '\n' => tokens.push(Token::NewLine),

            '.' => tokens.push(Token::Dot),
            '!' => tokens.push(Token::Bang),
            '=' => tokens.push(Token::Equal),
            ',' => tokens.push(Token::Comma),
            ' ' => tokens.push(Token::Space),

            '[' => tokens.push(Token::OpenBracket),
            ']' => tokens.push(Token::CloseBracket),
            '(' => tokens.push(Token::OpenParenthesis),
            ')' => tokens.push(Token::CloseParenthesis),

            '\'' => tokens.push(Token::SingleQuote),
            '"' => tokens.push(Token::DoubleQuote),

            '<' => tokens.push(Token::LessThan),
            '>' => tokens.push(Token::GreaterThan),

            '+' => tokens.push(Token::Plus),
            '-' => tokens.push(Token::Minus),
            '/' => tokens.push(Token::Divide),
            '*' => tokens.push(Token::Multiply),

            '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => {
                // Here we remove 48 when converting to an u8, to keep the numerical value
                // (instead of getting the unicode value)

                if let Some(Token::Litteral(inner)) = tokens.last_mut() {
                    inner.push(c);
                } else if let Some(Token::Numeric(inner)) = tokens.last_mut() {
                    inner.push(c as u8 - 48);
                } else {
                    tokens.push(Token::Numeric(vec![c as u8 - 48]));
                }
            }

            _ => {
                if let Some(Token::Litteral(inner)) = tokens.last_mut() {
                    inner.push(c);
                } else {
                    tokens.push(Token::Litteral(c.to_string()))
                }
            }
        }
    }

    Ok(tokens)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_single_character_tokens() {
        assert_debug_snapshot!(scan_one("."));
        assert_debug_snapshot!(scan_one("!"));
        assert_debug_snapshot!(scan_one("="));
        assert_debug_snapshot!(scan_one(","));
        assert_debug_snapshot!(scan_one("["));
        assert_debug_snapshot!(scan_one("]"));
        assert_debug_snapshot!(scan_one("("));
        assert_debug_snapshot!(scan_one(")"));
        assert_debug_snapshot!(scan_one("'"));
        assert_debug_snapshot!(scan_one("\""));
        assert_debug_snapshot!(scan_one("<"));
        assert_debug_snapshot!(scan_one(">"));
        assert_debug_snapshot!(scan_one("+"));
        assert_debug_snapshot!(scan_one("-"));
        assert_debug_snapshot!(scan_one("/"));
        assert_debug_snapshot!(scan_one("*"));
    }

    #[test]
    fn test_numeric_tokens() {
        assert_debug_snapshot!(scan_one("123"));
        assert_debug_snapshot!(scan_one("4567"));
        assert_debug_snapshot!(scan_one("0"));
    }

    #[test]
    fn test_literal_tokens() {
        assert_debug_snapshot!(scan_one("hello"));
        assert_debug_snapshot!(scan_one("world"));
        assert_debug_snapshot!(scan_one("foo bar"));
    }

    #[test]
    fn test_mixed_tokens() {
        assert_debug_snapshot!(scan_one("const var scores = [3, 2, 5]!"));
        assert_debug_snapshot!(scan_one("const x = 10 + 20!"));
        assert_debug_snapshot!(scan_one("if (x < 5) { return! }"));
    }

    #[test]
    fn test_edge_cases() {
        assert_debug_snapshot!(scan_one(""));

        assert_debug_snapshot!(scan_one("   "));
        assert_debug_snapshot!(scan_one("\t\t\t"));
        assert_debug_snapshot!(scan_one("\n\n\n"));

        assert_debug_snapshot!(scan_one("@#$%^&*()"));

        assert_debug_snapshot!(scan_one(&"a".repeat(1000)));

        assert_debug_snapshot!(scan_one("!!!"));
        assert_debug_snapshot!(scan_one("..."));
        assert_debug_snapshot!(scan_one(",,,"));

        assert_debug_snapshot!(scan_one("123abc!@#"));

        assert_debug_snapshot!(scan_one("[(])"));
        assert_debug_snapshot!(scan_one("[{]}"));

        assert_debug_snapshot!(scan_one("\"Hello\\nWorld\""));
        
        assert_debug_snapshot!(scan_one("00123"));
        assert_debug_snapshot!(scan_one("0001"));

        assert_debug_snapshot!(scan_one("Hello\x00World"));

        assert_debug_snapshot!(scan_one("12345678901234567890"));
    }
}
