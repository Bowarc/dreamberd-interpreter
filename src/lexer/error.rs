#[derive(Debug, thiserror::Error)]
pub enum ParserError{
    #[error("Source file not found")]
    SourceFileNotFound
}
