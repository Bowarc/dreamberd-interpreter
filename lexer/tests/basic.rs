#[cfg(test)]
mod tests {
    use lexer::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_single_character_tokens() {
        assert_debug_snapshot!(scan_one("."));
        assert_debug_snapshot!(scan_one("!"));
        assert_debug_snapshot!(scan_one("="));
        assert_debug_snapshot!(scan_one(","));
        assert_debug_snapshot!(scan_one("["));
        assert_debug_snapshot!(scan_one("]"));
        assert_debug_snapshot!(scan_one("("));
        assert_debug_snapshot!(scan_one(")"));
        assert_debug_snapshot!(scan_one("'"));
        assert_debug_snapshot!(scan_one("\""));
        assert_debug_snapshot!(scan_one("<"));
        assert_debug_snapshot!(scan_one(">"));
        assert_debug_snapshot!(scan_one("+"));
        assert_debug_snapshot!(scan_one("-"));
        assert_debug_snapshot!(scan_one("/"));
        assert_debug_snapshot!(scan_one("*"));
    }

    #[test]
    fn test_numeric_tokens() {
        assert_debug_snapshot!(scan_one("123"));
        assert_debug_snapshot!(scan_one("4567"));
        assert_debug_snapshot!(scan_one("0"));
    }

    #[test]
    fn test_literal_tokens() {
        assert_debug_snapshot!(scan_one("hello"));
        assert_debug_snapshot!(scan_one("world"));
        assert_debug_snapshot!(scan_one("foo bar"));
    }

    #[test]
    fn test_mixed_tokens() {
        assert_debug_snapshot!(scan_one("const var scores = [3, 2, 5]!"));
        assert_debug_snapshot!(scan_one("const const x = 10 + 20!"));
        assert_debug_snapshot!(scan_one("if (x < 5) { return! }"));
    }

    #[test]
    fn test_edge_cases() {
        assert_debug_snapshot!(scan_one(""));

        assert_debug_snapshot!(scan_one("   "));
        assert_debug_snapshot!(scan_one("\t\t\t"));
        assert_debug_snapshot!(scan_one("\n\n\n"));

        assert_debug_snapshot!(scan_one("@#$%^&*()"));

        assert_debug_snapshot!(scan_one(&"a".repeat(1000)));

        assert_debug_snapshot!(scan_one("!!!"));
        assert_debug_snapshot!(scan_one("..."));
        assert_debug_snapshot!(scan_one(",,,"));

        assert_debug_snapshot!(scan_one("123abc!@#"));

        assert_debug_snapshot!(scan_one("[(])"));
        assert_debug_snapshot!(scan_one("[{]}"));

        assert_debug_snapshot!(scan_one("\"Hello\\nWorld\""));

        assert_debug_snapshot!(scan_one("00123"));
        assert_debug_snapshot!(scan_one("0001"));

        assert_debug_snapshot!(scan_one("Hello\x00World"));

        assert_debug_snapshot!(scan_one("12345678901234567890"));
    }
}
