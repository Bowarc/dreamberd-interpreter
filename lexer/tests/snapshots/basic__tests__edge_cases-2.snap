---
source: lexer/tests/basic.rs
expression: "scan_one(\"   \")"
---
Ok(
    [
        Space,
        Space,
        Space,
    ],
)
